### PR TITLE
[CARBONDATA-3858] Check CDC deltafiles count in the testcase

### PR DESCRIPTION
**Pull request by marchpure**
### Why is this PR needed?
 Current there is no deltafiles count check in the testcase, which shall be supplemented.
 
 ### What changes were proposed in this PR?
Check CDC deltafiles count in the testcase

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1445/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3171/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala [hunk 1]
```diff
@@ -51,6 +52,7 @@ import org.apache.carbondata.core.mutate.CarbonUpdateUtil
 import org.apache.carbondata.core.util.CarbonProperties
 import org.apache.carbondata.processing.loading.FailureCauses
 
+
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:55]: revert this

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala [hunk 2]
```diff
@@ -269,11 +271,10 @@ case class CarbonMergeDataSetCommand(
       new SparkCarbonFileFormat().prepareWrite(sparkSession, job,
         Map(), schema)
     val config = SparkSQLUtil.broadCastHadoopConf(sparkSession.sparkContext, job.getConfiguration)
-    (frame.rdd.coalesce(DistributionUtil.getConfiguredExecutors(sparkSession.sparkContext)).
-      mapPartitionsWithIndex { case (index, iter) =>
+    (frame.rdd.mapPartitionsWithIndex { case (index, iter) =>
         CarbonProperties.getInstance().addProperty(CarbonLoadOptionConstants
           .ENABLE_CARBON_LOAD_DIRECT_WRITE_TO_STORE_PATH, "true")
-        val confB = config.value.value
+        val confB = new Configuration(config.value.value)
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:277]: why this change required? `confB` is already of type `Configuration`

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala [hunk 3]
```diff
@@ -547,6 +558,7 @@ class MergeTestCase extends QueryTest with BeforeAndAfterAll {
     CarbonMergeDataSetCommand(dwSelframe,
       odsframe,
       MergeDataSetMatches(col("A.id").equalTo(col("B.id")), matches.toList)).run(sqlContext.sparkSession)
+
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala:561]: revert this change

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala [hunk 4]
```diff
@@ -702,10 +716,25 @@ class MergeTestCase extends QueryTest with BeforeAndAfterAll {
...
       whenMatched("B.deleted=true").
       delete().execute()
+    assert(getDeleteDeltaFileCount("target") == 1)
     checkAnswer(sql("select count(*) from target"), Seq(Row(3)))
     checkAnswer(sql("select * from target order by key"), Seq(Row("c", "200"), Row("d", "3"), Row("e", "100")))
   }
 
+  private def getDeleteDeltaFileCount(tableName: String): Int = {
+    val table = CarbonEnv.getCarbonTable(None, tableName)(sqlContext.sparkSession)
+    val path = table.getTablePath
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala:726]: take the path till segment dir, that is till `Part0`

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala [hunk 5]
```diff
@@ -702,10 +716,25 @@ class MergeTestCase extends QueryTest with BeforeAndAfterAll {
...
+    val deleteDeltaFiles = FileFactory.getCarbonFile(path).listFiles(true, new CarbonFileFilter {
+      override def accept(file: CarbonFile): Boolean = file.getName.endsWith(CarbonCommonConstants
+        .DELETE_DELTA_FILE_EXT)
+    })
+    if (deleteDeltaFiles != null) {
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala:731]: no need of null check, it returns an array, so directly return `deleteDeltaFiles .size`

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala (hunk 1 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala]: modified

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala (hunk 2 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:275]: In multi-concurrent scenarios, conf is tampered, leading to some exceptions. A new configuration can solved this issue.

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala (hunk 3 above)
Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala]: modified

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala (hunk 5 above)
Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala]: modified

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala (hunk 4 above)
Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala:725]: I scan the dir with recursive = true. It will get # of all deltafiles.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3173/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1448/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala (hunk 2 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:275]: sorry, i did not understand conf is tampered, can you please tell what type error you got during concurrent scenarios?

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala (hunk 4 above)
Comment (akashrn5) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala:725]: yeah, saw that, but if we give the path till Part directory, we can reduce the files in list, so better to change the path.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala (hunk 2 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:275]: Without this change. The CI throws the exceptions (checked in serveral different envs):
java.lang.RuntimeException: Store location not set for the key __temptable-f8c5bbbf-0b73-4288-9438-146283d442c0_1592586283728_null_e8112868-3cf4-442d-bf9c-05f90bfca8240x0
	at org.apache.carbondata.processing.loading.TableProcessingOperations.deleteLocalDataLoadFolderLocation(TableProcessingOperations.java:125)


The root cause is 
the line: "val context = new TaskAttemptContextImpl(**confB**, attemptID)"
different context will have the same taskno, which is strange.

With this line, the exception disappears, CI passed.

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala (hunk 4 above)
Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala]: modified

Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/merge/MergeTestCase.scala]: Agree with you. Thanks a lot for this good find.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3174/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1449/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala (hunk 2 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:275]: i think adding new conf for it is not correct we need to analyze properly, may be you can revert these changes and we can handle during other cdc  optimizations

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3188/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1462/

**Comment by akashrn5** [comment]
LGTM

**Comment by xubo245** [comment]
LGTM, Thanks for your contribution!
